drm: rp1: rp1-dsi: Switch to PLL_SYS source for DPI when 8*lanes > bpp

njhollinghurst opened the pull request:
Draft which partially addresses #6125.

This sets up the DPI clock as accurately as it can to match the DSI rate; without altering HS/LP transitions or messing with the mode timings.

If 4-lane panels are as fussy as the bridge chip in our (1-lane) 7" display, there is a risk of horizontal jitter, due to irregular line intervals at most pixel-clock rates.

Not yet successfully tested with any real 4-lane panel.

njhollinghurst committed:
drm: rp1: rp1-dsi: Switch to PLL_SYS source for DPI when 8*lanes > bpp

To support 4 lanes, reparent DPI clock source between DSI byte-clock
(which is declared to run at a dummy rate) and PLL_SYS (at real 200MHz).
This requires DSI to know about PLL_SYS; updated this in rp1.dtsi.

Tidying: All frequencies now in Hz (not kHz), where DSI speed is now
represented by byte_clock to simplify arithmetic. Clamp DPI and byte
clocks to their legal ranges; remove the lower bound on HSTX timeout.
Change the "dummy" byte-clock frequency from 72MHz to 72kHz, to more
easily distinguish it from real clock frequencies.

The resulting DSI output appears largely correct, but line timings
become slightly irregular at most pixel rates. For reasons not yet
understood, RP1 DSI still doesn't work with LT070ME05000.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

6by9 commented on drivers/gpu/drm/rp1/rp1-dsi/rp1_dsi_dsi.c:
```diff
@@ -1124,19 +1124,23 @@ static uint8_t dphy_get_div(u32 refclk_khz, u32 vco_freq_khz, u32 *ptr_m, u32 *p
 
-	for (n = 1 + refclk_khz / REF_DIVN_MAX; n * REF_DIVN_MIN <= refclk_khz && n < 100; ++n) {
-		u32 half_m = (n * vco_freq_khz + refclk_khz) / (2 * refclk_khz);
+	for (n = 1 + refclk / REF_DIVN_MAX; n * REF_DIVN_MIN <= refclk && n < 100; ++n) {
```
Need to be careful to ensure this can't be built on a 32bit kernel as it'll error out on 64bit division. I'd hope a Kconfig change can protect against that.

njhollinghurst replied:
Can a 32-bit kernel not do u64 division? I can probably re-work the arithmetic to avoid it but it would look messier.

6by9 replied:
It can, but not as a raw divide.

`do_div`, or you've got `DIV_ROUND_CLOSEST`, `DIV_ROUND_DOWN`, and `DIV_ROUND_UP` type macros (see include/linux/math.h)

pelwell replied:
I'm sure it can - the compiler will generate calls to helper functions - provided at least one of the types is 64-bit.

pelwell replied:
The kernel build checks have all been successful.

njhollinghurst replied:
`div_u64()` looks like the one to use.

6by9 replied:
CONFIG_DRM_RP1_DSI is only set in arch/arm64/configs/bcm2711_defconfig and arch/arm64/configs/bcm2712_defconfig, so no 32bit build is trying to compile it.

If it were to try, then it typically fails with errors such as 
```
ERROR: "__aeabi_ldivmod" [/home/pi/media_build/v4l/mn88436.ko] undefined!
ERROR: "__aeabi_uldivmod" [/home/pi/media_build/v4l/mn88436.ko] undefined!
```
Certainly for upstreaming you'll get the random config builds managing to try building it on 32bit and failing. Kconfig should block that.

My 32bit build + MFD_RP1 and DRM_RP1_DSI has just failed with 
```
ERROR: modpost: "msi_domain_set_affinity" [drivers/mfd/rp1.ko] undefined!
ERROR: modpost: "irq_domain_deactivate_irq" [drivers/mfd/rp1.ko] undefined!
ERROR: modpost: "irq_domain_activate_irq" [drivers/mfd/rp1.ko] undefined!
```
instead, so bigger fish to fry.

More an observation than requiring fixing immediately.

pelwell replied:
(Largely to myself)
Apparently this is because the kernel doesn't use libgcc and doesn't include its own versions of these functions. The functions declared in math64.h should be used instead.

njhollinghurst replied:
I think the u64 divides are all fixed up now. refclk is back to being a u32.

6by9 commented on drivers/gpu/drm/rp1/rp1-dsi/rp1_dsi.c:
```diff
@@ -54,11 +54,18 @@ static void rp1_dsi_bridge_pre_enable(struct drm_bridge *bridge,
+	if (!dsi->dsi_running) {
+		rp1dsi_dsi_setup(dsi, &dsi->pipe.crtc.state->adjusted_mode);
+		dsi->dsi_running = true;
+	}
```
How are you thinking you can get to `enable` without `pre_enable` being called?

njhollinghurst replied:
Ah, I wasn't sure if it was mandatory. BTW it would be nice if pre-enable was called before any DSI commands got sent.

njhollinghurst replied:
Gone.

6by9 commented on drivers/gpu/drm/rp1/rp1-dsi/rp1_dsi_dsi.c:
```diff
@@ -1290,23 +1300,25 @@ static unsigned long rp1dsi_refclk_freq(struct rp1_dsi *dsi)
+		return;
+
+	if (bpp >= 8 * lanes && dsi->clocks[RP1DSI_CLOCK_BYTE] &&
+	    clk_set_parent(dsi->clocks[RP1DSI_CLOCK_DPI], dsi->clocks[RP1DSI_CLOCK_BYTE]) == 0)
```
`!clk_set_parent(...)`

njhollinghurst replied:
"not" sounds like failure. Is that the preferred style?

pelwell replied:
Yes, sadly

6by9 replied:
You're in functions that return void, so you cant actually return an error even if we could handle failing to set up the clock tree.

The normal preference is to use `!foo` rather than `foo == 0`, but it doesn't appear to be in https://www.kernel.org/doc/html/latest/process/coding-style.html

6by9 commented on drivers/gpu/drm/rp1/rp1-dsi/rp1_dsi_dsi.c:
```diff
@@ -1290,23 +1300,25 @@ static unsigned long rp1dsi_refclk_freq(struct rp1_dsi *dsi)
+	    clk_set_parent(dsi->clocks[RP1DSI_CLOCK_DPI], dsi->clocks[RP1DSI_CLOCK_BYTE]) == 0)
+		byte_clock = clk_get_rate(dsi->clocks[RP1DSI_CLOCK_BYTE]); /* dummy/nominal rate */
+	else if (dsi->clocks[RP1DSI_CLOCK_PLLSYS])
+		clk_set_parent(dsi->clocks[RP1DSI_CLOCK_DPI], dsi->clocks[RP1DSI_CLOCK_PLLSYS]);
```
You checked the return code above, but not here.

njhollinghurst replied:
Yes, not really sure what to do if they both fail, as it leaves us with an unknown clock parent (which might still work provided it has never been the dummy one).

6by9 replied:
Yes, we're in void returning functions, so you are pretty stuffed. I was more commenting on the lack of consistency.

Doing a `git grep clk_set_parent` I see a number of drivers not checking the return value at all, so it's probably OK not to.

njhollinghurst replied:
Now checking neither of them.

6by9 commented on drivers/gpu/drm/rp1/rp1-dsi/rp1_dsi_dsi.c:
```diff
@@ -1337,17 +1349,17 @@ static u32 get_colorcode(enum mipi_dsi_pixel_format fmt)
-	int lane_kbps;
 	unsigned int bpp = mipi_dsi_pixel_format_to_bpp(dsi->display_format);
+	u32 byte_clock = clamp((bpp * 125u * (u32)min(mode->clock, 200000)) / dsi->lanes,
+			       10000000u, 187500000u);
```
Magic numbers,

njhollinghurst replied:
Max DPI pixel rate in kHz; min and max DSI rates in bytes/second. I can add them to the #defines above.

6by9 replied:
I would, just to make it clear.

njhollinghurst replied:
Done.

6by9 reviewed (commented):
I'll trust you on the PLL setup. Largely coding style nits.

njhollinghurst commented:
> I'll trust you on the PLL setup. Largely coding style nits.

The "bogus fixed clock" is also something I really ought to fix -- probably by making a near clone of `clk_fixed`  whose rate can notionally be changed, since it's really not a RP1-specific thing -- but I'll try to get everything functional first.

njhollinghurst commented:
Hmm, doing the clocks "correctly" is not that simple. The current dummy frequency of 72kHz was chosen to be highly divisible; this is not true of an an actual MIPI VCO rate rounded to integer Hz. And once rounding errors creep in we might not be able to guarantee that the divider will be the precise integer or simple fraction that we wanted it to be, or that `clk-rp1` will consistently favour an exact ratio over an inexact one.

njhollinghurst committed:
clk, drm: rp1-dsi: Add "varsrc" clocks to represent MIPI byte clocks

Add a new class of clocks to RP1 to represent clock sources whose
frequencies change at run-time as a side-effect of some other driver.
Remove the "dummy" 72kHz clock sources from rp1.dtsi. Set the actual
MIPI byte clock frequency from RP1 DSI driver. (These changes should
be made atomically).

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

njhollinghurst commented:
I've left the clocking changes unsquashed, for clarity. There are changes in several places which kinda need to be atomic.
@naushir do you have any comments on the clk-rp1 changes?

naushir commented:
> @naushir do you have any comments on the clk-rp1 changes?

clk-rp1 looks reasonable to me.  Shame there is not an existing clock type like this in the framework.

njhollinghurst committed:
drm: rp1: rp1-dsi: Fix up HSTX TO (again); optionally stay in HS for HFP

Set HSTX timeout to the whole vertical active region, not just one line.
For this, the divider must be increased, which affects the LPRX timeout
(but not the BTA one). Disable the timeout altogether when too long.

Disable returning to LP state during HBP, when sync pulses are required
and we can't do accurate DPI timing. This should guarantee that the
Hsync->Data interval is a fixed number of DSI clocks in those cases.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

njhollinghurst commented:
Replaced by #6157
